flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -426,6 +426,93 @@ static bool hasPointerType(mlir::Type type) {
 }
 
+// This is a NF performance hack. It makes a simple test that the slices of the
+// load, \p ld, and the merge store, \p st, are trivially mutually exclusive.
+static bool mutuallyExclusiveSliceRange(ArrayLoadOp ld, ArrayMergeStoreOp st) {
+  // If the same array_load, then no further testing is warranted.
+  if (ld.getResult() == st.original())
+    return false;
+
+  auto getSliceOp = [](mlir::Value val) -> fir::SliceOp {
+    if (!val)
+      return {};
+    auto sliceOp = mlir::dyn_cast_or_null<fir::SliceOp>(val.getDefiningOp());
+    if (!sliceOp)
+      return {};
+    return sliceOp;
+  };
+
+  auto ldSlice = getSliceOp(ld.slice());
+  auto stSlice = getSliceOp(st.slice());
+  if (!ldSlice || !stSlice)
+    return false;
+
+  // Resign on subobject slices.
+  if (!ldSlice.fields().empty() || !stSlice.fields().empty() ||
+      !ldSlice.substr().empty() || !stSlice.substr().empty())
+    return false;
+
+  // Crudely test that the two slices do not overlap by looking for the
+  // following general condition. If the slices look like (i:j) and (j+1:k) then
+  // these ranges do not overlap. The addend must be a constant.
+  auto ldTrips = ldSlice.triples();
+  auto stTrips = stSlice.triples();
+  const auto size = ldTrips.size();
+  if (size != stTrips.size())
+    return false;
+
+  auto displacedByConstant = [](mlir::Value v1, mlir::Value v2) {
+    auto removeConvert = [](mlir::Value v) -> mlir::Operation * {
+      auto *op = v.getDefiningOp();
+      while (auto conv = mlir::dyn_cast_or_null<fir::ConvertOp>(op))
+        op = conv.value().getDefiningOp();
+      return op;
+    };
+
+    auto isPositiveConstant = [](mlir::Value v) -> bool {
+      if (auto conOp =
+              mlir::dyn_cast<mlir::arith::ConstantOp>(v.getDefiningOp()))
+        if (auto iattr = conOp.value().dyn_cast<mlir::IntegerAttr>())
+          return iattr.getInt() > 0;
+      return false;
+    };
+
+    auto *op1 = removeConvert(v1);
+    auto *op2 = removeConvert(v2);
+    if (!op1 || !op2)
+      return false;
+    if (auto addi = mlir::dyn_cast<mlir::arith::AddIOp>(op2))
+      if ((addi.lhs().getDefiningOp() == op1 &&
+           isPositiveConstant(addi.rhs())) ||
+          (addi.rhs().getDefiningOp() == op1 && isPositiveConstant(addi.lhs())))
+        return true;
+    if (auto subi = mlir::dyn_cast<mlir::arith::SubIOp>(op1))
+      if (subi.lhs().getDefiningOp() == op2 && isPositiveConstant(subi.rhs()))
+        return true;
+    return false;
+  };
+
+  for (std::remove_const_t<decltype(size)> i = 0; i < size; i += 3) {
+    // If identical, skip to the next triple.
+    if (ldTrips[i] == stTrips[i] && ldTrips[i + 1] == stTrips[i + 1] &&
+        ldTrips[i + 2] == stTrips[i + 2])
+      continue;
+    // If both are loop invariant, skip to the next triple.
+    if (mlir::isa_and_nonnull<fir::UndefOp>(ldTrips[i + 1].getDefiningOp()) &&
+        mlir::isa_and_nonnull<fir::UndefOp>(stTrips[i + 1].getDefiningOp()))
+      continue;
```

Comment:
If both are loop single index constants that are different, there is also trivially no overlap. e.g. `x(1, :) = x(2, :)` does not overlap, that may be an easy catch here (OK if you leave it outside though given I understand this is not a long term solution here).

Reply:
Solving distinct invariants wasn't the goal here. Not saying it can't be done.